pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,269 @@
+				glog.V(Vdebug).Infof("processPodItem: CloneOf annotatated in pvc %q", pvcKey)
+			}
+		}
+
+	}
+	return nil
+}
+
+// Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
+// We forget the key unless `processPvcItem` returns an error in which case the key can be
+// retried.
+func (c *CloneController) ProcessNextPvcItem() bool {
+	key, shutdown := c.pvcQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.pvcQueue.Done(key)
+
+	pvc, err := c.pvcFromKey(key)
+	if err != nil || pvc == nil {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: error converting key %v to pvc: %v", key, err))
+	}
+	// filter pvc and decide if the source and target pods should be created
+	if continue_processing, _, _ := checkClonePVC(c.clientset, pvc, false); !continue_processing {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: skipping pvc %q\n", key))
+	}
+	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
+	err = c.processPvcItem(pvc)
+	if err != nil { // processPvcItem errors may not have been logged so log here
+		glog.Errorf("error processing pvc %q: %v", key, err)
+		return true
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
+}
+
+// Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
+// processing this pvc, which would result in multiple pods for the same pvc.
+func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
+	cr, err := getCloneRequestPVC(pvc)
+	if err != nil {
+		return err
+	}
+
+	// check our existing pvc one more time to ensure we should be working on it
+	// and to help mitigate any race conditions. This time we get the latest pvc.
+	doCreate, pvc, err := checkClonePVC(c.clientset, pvc, true)
+	if err != nil { // maybe an intermittent api error
+		return err
+	}
+	if !doCreate { // don't create pods but not an error
+		return nil // forget key; logging already done
+	}
+
+	// update pvc with 'AnnCloningPods' annotation to indicate cloning is in process
+	anno := map[string]string{AnnCloningPods: "exist"}
+	var lab map[string]string
+	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
+	if err != nil {
+		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
+	}
+
+	//create random string to be used for pod labeling and hostpath name
+	generatedLabelStr := GenerateLabelStr(GENERATED_CLONING_LABEL_LEN)
+	//create the source pod
+	_, err = CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
+	if err != nil {
+		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
+		return err
+	}
+
+	//create the target pod
+	_, err = CreateCloneTargetPod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
+	if err != nil {
+		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
+		return err
```

Comment:
If the TODO is implemented then we have resiliency against intermittent api failures, which is good and needed. However, given that's done, if the source pod is created but the target pod Create fails **and** the anno is removed, there's a problem that the source pod will be created again.